### DISCUSSION TITLE
Can We Use Tools with GroupChat?

### DISCUSSION THREAD

**Discussion by kiranraaj19** [microsoft/autogen#4488] [Q&A] [answered]
I have been trying to Integrate Tools into agents that are in a groupchat.

I have a groupchat defined of 3 Assistant agents. And I tried registering function for the user proxy and one of the agent in the groupchat.

Code Snippet:
```
# manager is defined properly with Custom Agents inherited from AssistantAgents
agentchat.register_function(
                    download_code_tool,
                    caller = manager.groupchat.agents[-1],
                    executor = user_proxy,
                    name="download_code_tool",
                    description="Tool used for saving code files to Downloads folder"
                )

# User proxy is properly defined
user_proxy.initiate_chat(
        manager,
        message=input_msg
    )
```

But I keep getting these errors:
```
ValueError: No agent can execute the function download_code_tool. Please check the function_map of the agents.
```

The Chat happens properly till we reach the particular agent i registered it with.

I have registered the function to user_proxy and one of the agent in manager following the docs provided in Autogen Documentation. [Documentation that i followed](https://microsoft.github.io/autogen/0.2/docs/tutorial/tool-use/)

Unfortunately, I couldnt find examples where tools was used in groupchat context, The examples i found too, They used @ decorators to register, So I went along with agentchat.register_function because i found that best for modular code.

I even tried searching for alternatives, Came across ToolAssistantAgents but those are deprecated. I see that there is no question on tool usage along with groupchat, So maybe I am doing something wrong? Can someone guide me.

**Comment by ekzhu** [comment, marked as answer]
You need to make sure the caller is not user proxy. 

If you are starting new, I recommend starting with the v0.4 AgentChat API, and take a look at the Teams available. https://microsoft.github.io/autogen/dev/user-guide/agentchat-user-guide/index.html

> Came across ToolAssistantAgents but those are deprecated.

In v0.4, it is AssistantAgent.

**Reply by kiranraaj19** [reply]
Hey, Thanks for the help. I realized i was importing Agents from autogen rather than autogen_agentchat and its way simpler to develop Agents with the tools parameter now. But I still have a problem while following this [documentation](https://microsoft.github.io/autogen/dev/user-guide/agentchat-user-guide/tutorial/selector-group-chat.html)

I have setup agents and SelectorGroupChat properly but i get this following error:

```
in _init
    await ClosureAgent.register(
          ^^^^^^^^^^^^^^^^^^^^^^
TypeError: BaseAgent.register() got an unexpected keyword argument 'closure'
```

I am not using ClosureAgent anywhere!

Code I am using:
```
# agents are defined according to documentation
team = SelectorGroupChat(
        agents,
        model_client=az_model_client,
        termination_condition=termination
    )

stream = team.run_stream(task = input_message)
async for message in stream:
    print(message)
```

Is there something that I am missing?

**Reply by ekzhu** [reply]
Which version are you using? And your Python version? 

Also be aware you need to use the documentation version that matches the dev build version. The code is currently under development.

**Reply by kiranraaj19** [reply]
Hey, Clean installing the libraries worked, Thankyou so much for your help!
